crates/server/src/server/response_handler.rs

```diff
@@ -46,12 +47,44 @@ pub trait ResponseHandler: Clone + Send + Sync + Unpin + 'static {
 pub struct ResponseHandle {
     dst: SocketAddr,
     stream_handle: BufDnsStreamHandle,
+    protocol: Protocol,
 }
 
 impl ResponseHandle {
     /// Returns a new `ResponseHandle` for sending a response message
-    pub fn new(dst: SocketAddr, stream_handle: BufDnsStreamHandle) -> Self {
-        Self { dst, stream_handle }
+    pub fn new(dst: SocketAddr, stream_handle: BufDnsStreamHandle, protocol: Protocol) -> Self {
+        Self {
+            dst,
+            stream_handle,
+            protocol,
+        }
+    }
+
+    /// Selects an appropriate maximum serialized size for the given response.
+    fn max_size_for_response<'a>(
+        &self,
+        response: &MessageResponse<
+            '_,
+            'a,
+            impl Iterator<Item = &'a Record> + Send + 'a,
+            impl Iterator<Item = &'a Record> + Send + 'a,
+            impl Iterator<Item = &'a Record> + Send + 'a,
+            impl Iterator<Item = &'a Record> + Send + 'a,
+        >,
+    ) -> u16 {
+        // Use EDNS, if available.
+        if let Some(edns) = response.get_edns() {
+            edns.max_payload()
+        } else {
+            // No EDNS. Use an appropriate maximum for the protocol.
+            match self.protocol {
+                Protocol::Udp => {
+                    // For UDP, we use the recommended max from RFC6891.
+                    4096
```

Comment:
Should this reference the static in `trust_dns_proto::udp::MAX_RECEIVE_BUFFER_SIZE` that you added instead?